projects/driver_controls/src/stm32f0xx/ads1015.c

```diff
@@ -0,0 +1,171 @@
+// pin is asserted when a conversion result is ready). Ads1015_init registers prv_interrupt_handler
+// as the callback for this pin. Once the interrupt is raised, it means that a conversion has been
+// completed. prv_interrupt_handler reads the conversion values from the registers and stores them
+// in storage, and then switches the channel to the next enabled one.
+// The user could access conversion data by the provided read functions. These
+// functions get their data from storage and not directly by reading the registers. Also look at
+// ads1015_def.h for more information on macros and how registers are setup.
+// Reading and writing to registers is done through I2C interface. Look at section 8.5.3 of the
+// datasheet for the timing diagrams.
+#include <status.h>
+#include <stdio.h>
+#include <string.h>
+#include "ads1015_def.h"
+#include "gpio_it.h"
+
+// Updates the channel_enable bitset when a channel is enabled/disabled.
+static void prv_mark_channel_enabled(Ads1015Channel channel, bool enable, uint8_t *channel_enable) {
+  if (enable) {
+    *channel_enable |= 1 << channel;
+  } else {
+    *channel_enable &= ~(1 << channel);
+  }
+}
+
+// Writes to register given upper and lower bytes.
+static StatusCode prv_setup_register(Ads1015Storage *storage, uint8_t reg, uint8_t msb,
+                                     uint8_t lsb) {
+  uint8_t ads1015_setup_register[] = { reg, msb, lsb };
+  return i2c_write(storage->i2c_port, storage->i2c_addr + ADS1015_I2C_BASE_ADDRESS,
+                   ads1015_setup_register, SIZEOF_ARRAY(ads1015_setup_register));
+}
+
+// Reads the register and stores the value in the given array.
+static StatusCode prv_read_register(I2CPort i2c_port, Ads1015Address i2c_addr, uint8_t reg,
+                                    uint8_t *rx_data, size_t rx_len) {
+  status_ok_or_return(i2c_write(i2c_port, i2c_addr + ADS1015_I2C_BASE_ADDRESS, &reg, sizeof(reg)));
+
+  status_ok_or_return(i2c_read(i2c_port, i2c_addr + ADS1015_I2C_BASE_ADDRESS, rx_data, rx_len));
+  return STATUS_CODE_OK;
+}
+
+// Switches to the given channel by writing to config register.
+static StatusCode prv_set_channel(Ads1015Storage *storage, Ads1015Channel channel) {
+  if (channel >= NUM_ADS1015_CHANNELS) {
+    return status_code(STATUS_CODE_INVALID_ARGS);
+  }
+  status_ok_or_return(prv_setup_register(storage, ADS1015_ADDRESS_POINTER_CONFIG,
+                                         ADS1015_CONFIG_REGISTER_MSB(channel),
+                                         ADS1015_CONFIG_REGISTER_LSB));
+  storage->current_channel = channel;
+  return STATUS_CODE_OK;
+}
+
+// This function is registered as the callback for ALRT/RDY Pin.
+// Reads and stores the conversion value in storage, and switches to the next enabled channel.
+// Also if there is a callback on a channel, it will be run here.
+static void prv_interrupt_handler(const GPIOAddress *address, void *context) {
```

Comment:
Can you just add some newlines between blocks to help break up the logical sections? It's just a little difficult to read because it's just a massive block of code.